### PR TITLE
WorldQuery::Config for dynamic queries

### PR DESCRIPTION
**Pull request by Suficio**
# Objective

Allows providing runtime state to QueryState allowing users to implement dynamic queries which cannot be derived from type definition.

This PR is a cut-down and rebased version of https://github.com/bevyengine/bevy/pull/6390.

https://github.com/bevyengine/bevy/pull/6390 is itself an alternative proposal to https://github.com/bevyengine/bevy/pull/6240 which I believe to be a simpler approach due to no type magic.

I am making this cut-down PR as neither of the prior PRs received much attention in the way of reviews but both provide a powerful feature.

## Solution

- introduce a `Config` associated type to `WorldQuery`
    - this is `Config = ()` for current query types `&T`, `&mut T`, `Q::Config` for `Option<Q>` and `(A::Config, B::Config, ..)` for `(A, B, ..)`
- Make `QueryState::new` require `Q::Config: Default` and `F::Config: Default`. This propagates quite far, to `world.query`, `Query: SystemParam` and `ExtractComponentPlugin`

### CODE REVIEW

**Comment by Suficio** [comment]
> > #6390 is itself an alternative proposal to #6240 which I believe to be a simpler approach due to no type magic.
> 
> With #6240, how would you implement `init_state` for untyped query parameters?
> 
> You can't do `Query::<&Transform>::new_with_state(other_component_id)` so you impl `WorldQuery for Ptr<'_>` and do `Query::<Ptr<'_>>::new_with_state(component_id)`. But nothing in the type system specifies that you can not use `new_with_state` for `&Transform` and you must use `new_with_state` for `Ptr<'_>`.
> 
> So the only option for `init_state` would be to panic.

That's right, it would panic, and from the outside, unless you chose to only use `WorldQuery` implementations you know won't panic, you have no way of telling them apart. I think ultimately that is a very strong aspect of this approach.

Overall this PR achieves the same goal so I would like to push to get it adopted :)

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -988,8 +995,8 @@ impl World {
         match self.get_resource() {
             Some(x) => x,
             None => panic!(
-                "Requested resource {} does not exist in the `World`. 
-                Did you forget to add it using `app.insert_resource` / `app.init_resource`? 
+                "Requested resource {} does not exist in the `World`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:998]: These error messages changes are unrelated: we should strip them from this PR.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 2]
```diff
@@ -328,6 +328,9 @@ pub unsafe trait WorldQuery {
     /// constructing [`Self::Fetch`](crate::query::WorldQuery::Fetch).
     type State: Send + Sync + Sized;
 
+    /// Runtime config value passed to [`WorldQuery::init_state`]
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:331]: This needs expansion for users. What is this type used for, and how should it be chosen?

File: crates/bevy_ecs/src/query/state.rs [hunk 3]
```diff
@@ -95,9 +99,22 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
...
     /// Creates a new [`QueryState`] from a given [`World`] and inherits the result of `world.id()`.
-    pub fn new(world: &mut World) -> Self {
-        let fetch_state = Q::init_state(world);
-        let filter_state = F::init_state(world);
+    pub fn new(world: &mut World) -> Self
+    where
+        <Q as WorldQuery>::Config: Default,
+        <F as WorldQuery>::Config: Default,
+    {
+        QueryState::new_with_config(world, Default::default(), Default::default())
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/state.rs:107]:
```suggestion
        let fetch_config = Default::default();
        let filter_config = Default::default();
        QueryState::new_with_config(world, fetch_config, filter_config)
```

Cosmetic, but I think this makes the code significantly easier to follow at a glance.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 4]
```diff
@@ -328,6 +328,9 @@ pub unsafe trait WorldQuery {
...
+    type Config;
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:332]: Is there any use case for this to *not* have a `Default` trait?

It seems like we could significantly clean up this PR by just swapping to a trait bound here.

Comment (Suficio) [reply crates/bevy_ecs/src/query/fetch.rs:332]: If you wanted to implement a dynamic query, you would want to pass as `Config = ComponentId` and `ComponentId` would not implement `Default`.

File: crates/bevy_ecs/src/query/fetch.rs (hunk 2 above)
Comment (Suficio) [reply crates/bevy_ecs/src/query/fetch.rs:331]: Done

File: crates/bevy_ecs/src/world/mod.rs (hunk 1 above)
Comment (Suficio) [reply crates/bevy_ecs/src/world/mod.rs]: Done, autoformatter snuck in :)

File: crates/bevy_ecs/src/query/fetch.rs [hunk 5]
```diff
@@ -328,6 +328,22 @@ pub unsafe trait WorldQuery {
...
+    ///
+    /// Statically defined queries for types such as `&T` may obtain a
+    /// [`ComponentId`] statically, however, a dynamic query would require that
+    /// [`ComponentId`] is passed at runtime, such that
+    /// `Self::Config = ComponentId`.
+    ///
+    /// [`QueryState::new`](crate::query::state::QueryState::new) can be used when
+    /// `Self::Config: Default` otherwise
+    /// [`QueryState::new_with_config`](crate::query::state::QueryState::new_with_config)
+    /// is used to construct [`QueryState`](crate::query::state::QueryState).
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:344]:
```suggestion
    /// This associated type is defines how a query can be configured at runtime.
    ///
    /// [`QueryState::new_with_config`](crate::query::state::QueryState::new_with_config) uses this
    /// associated type to construct [`QueryState`](crate::query::state::QueryState). 
    /// When this associated type implements the `Default` trait, 
    /// [`QueryState::new`](crate::query::state::QueryState::new) can be used instead.
    /// 
    /// This is the case for statically defined queries for types such as `&T`, which can obtain a
    /// [`ComponentId`] statically.
    /// In the case where no config is required, `()` is conventionally chosen as the type for `Config`.
    ///
    /// However, the [`ComponentId`]s used by dynamic queries vary at runtime.
    /// By supplying `Config = ComponentId` we can vary the set of components requested
    /// each time we construct a new query, even for the same underlying Rust type.
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Thanks for the extended explanation: I think I have a good sense of how this is supposed to be used now :)

I've left some improved docs: feel free to workshop them as you see fit.

File: crates/bevy_ecs/src/query/fetch.rs (hunk 5 above)
Comment (Suficio) [reply crates/bevy_ecs/src/query/fetch.rs]: Finalized the doc comment :)

**Review by alice-i-cecile — APPROVED** [review]
I'm happy to move forward with this as an initial step now.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 6]
```diff
@@ -328,6 +328,24 @@ pub unsafe trait WorldQuery {
...
+    ///
+    /// In the case where no config is required, `()` is conventionally chosen
+    /// as the type for `Config`. This is the case for statically defined
+    /// queries, for types such as `&T`, since they can obtain a
+    /// [`ComponentId`] without any runtime information.
+    ///
+    /// However, for dynamic queries, the [`ComponentId`] varies at runtime and
+    /// mut be additionally provided. By supplying `Config = ComponentId` we
+    /// can vary the component requested each time we construct a new query,
+    /// even for the same underlying Rust type.
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/query/fetch.rs:346]: I'm trying to wrap my head around how this works but having a difficult time. Would it be worth adding a test or something so people like me can visualize it in action? 😅

A test would also allow us to verify this is working as intended for these dynamic usages.

Comment (Suficio) [reply crates/bevy_ecs/src/query/fetch.rs:346]: I don't have a test in this PR as it would require implementing a dynamic query from scratch which would be quite large.

@jakobhellermann has multiple example implementations, `Ptr, MutPtr, Vec`, in their PR:
https://github.com/jakobhellermann/bevy/blob/5617e3bb8337ae79e25445d468d274ec962d0305/crates/bevy_ecs/src/query/fetch.rs#L1546

I purposefully didn't include these implementations in this PR as it's not clear how to implement them in the "best" way, however, this PR allows users to at least implement them themselves :)

**Comment by Suficio** [comment]
Resolved by #9774
